### ISSUE TITLE
Fix/improve the callbacks for DataState especially with errors. 

### ISSUE THREAD

**Issue by levibostian** [levibostian/Driver-Android#2]
Let's say you are observing (via rxjava) a `DataState` object in your UI of your Android app. You probably have code that looks something like this:

```
        reposViewModel.observeRepos()
                .observe(this, Observer { reposState ->
                    reposState?.deliver(object : OnlineDataStateListener<List<RepoModel>> {
                        override fun firstFetchOfData() {
                             // handle me
                        }
                        override fun isEmpty() {
                             // handle me
                        }
                        override fun data(data: List<RepoModel>, fetched: Date) {
                             // handle me
                        }
                        override fun error(error: Throwable) {
                            // handle me
                        }
                        override fun fetchingFreshData() {
                           // handle me
                        }
                        override fun finishedFetchingFreshData(errorDuringFetch: Throwable?) {
                            // handle me
                        }
                    })
                })
```

Here is the current state of this code as of [this commit](https://github.com/levibostian/Teller-Android/commit/66b1601df35959cfe31c5938e0c836916c68084b):

This code above does work as expected (all callbacks are called as you want them to. Callbacks get called when changes happen. It's dependable) but only on one condition: For the error state, you are showing the user the error message of the error using something like a dialog. An independent UI that is not shared between any of the other states. 

That last part is very important and it explains the issue here. As long as you are handling the error in an independent UI, you will not see a problem in your app's code. This code above works and it works well. 

However, what if you want to share a UI with all of the states of your data? What if you have a textview in your UI that will say, "Empty. No data here.", "Loading data for the first time", "You have an error", "Here is your data". This seems like a reasonable thing you may want to do. With the current state of this library, you cannot do this in a dependable way. Why? Because when an error occurs in an `OnlineRepository` subclass and there is an error that happens (tested in the fetch function) such as a 404 or something like that and you return a `FetchResponse.fail()` result for the fetch, `error()` *will* be called in the code pasted above, but `empty()` will also be called immediately after. So if you want to share a UI with all of the states of your data, you will show a message "You have an error" and then a split second later you will switch to "Empty. No data here."

I want to improve this API. I want you to be able to know the exact state of your data and your UI can depend on Teller giving it the data it needs to show the state of that data to your user. It needs to be more dependable and not have any rules like I do here by saying error() needs to be handled in it's own UI. 

I believe to have this happen, I cannot simply edit the implementation of `OnlineRepository`. I believe it will take a refactor of the DataState deliver function as well. Perhaps a brand new set of callbacks, a brand new API to receive callbacks, I am not sure. Below I will share ideas and progress on this.

**Comment by levibostian** [comment]
Now that I think about it, this issue might be fixed with a refactor of `OnlineRepository`. I say this because `OnlineDataState` *does* already contain all of the different states that data can be in:

![img_20180517_140807](https://user-images.githubusercontent.com/2041082/40198832-2bc8b6f2-59dd-11e8-81a5-18c693619f10.jpg)

At it's current state, `OnlineDataState` does contain all of these different states in it. 

---

Possible solutions:

The original post above, that scenario is mostly about 1 specific use case. When you fetch data for the very first time and you have a network error (human caused like a 400 error or not user caused such as network timeout) while fetching data over the network. Right now, `OnlineRepository` will begin observing the cached data no matter if the fetch was successful or not. Something I should debug is commenting out the initial starting of observing data when data is first fetched and not successful to see if this problem above happens (empty() being called after error()). 

I do want to continue observing data changes. I don't want to try a possible solution of manually calling `getCachedData` during certain moments when I can instead be observing data. Changes should only happen when the app manually makes a change to data (without Teller's knowledge) or when a fetch call was successful and Teller calls `saveData()` on the `OnlineRepository` subclass (no matter if this was the first or not first time fetching data for that particular piece of data). 

Even though I do want to continue observing data changes, I believe I can be more smart and opinionated about when observing data happens. Right now, the `OnlineRepository` is pretty much always observing the cached data. Errors or no errors, it does not matter. I can have the `OnlineRepository` only observe cached data either if cached data already exists for that piece of data on first start up of this `OnlineRepository` subclass, or a first fetch was successful. 

But also, I have 2 more thoughts. 1. Errors are important and we should/can be more opinionated about them. When errors happen in your app, your user should be alerted, right? Maybe in the documentation of Teller, I mention that the `OnlineDataState` `error()` callback *should* have an independent UI to show that error. Or you should take that error, display the message to your user in the app, force them to acknowledge that it happened, and then we can display the other state of your data. Teller could have an API where when errors happen, they are sticky in `OnlineDataState`. They will not stop calling the `error()` callback until the app marks that error as resolved. So if `OnlineDataState` ever receives another callback via `deliver`, we check if there is an error, deliver all of the other states and then call `error()` on that sticky error last (to make sure the UI catches the error state last for changes) and continue calling it until it's marked as resolved. 

This is a great idea as it will make sure that errors do not go unnoticed by accident. However, there are a couple issues:

1. Dev forgetting to mark an error as resolved while developing. 
2. What if there is simply a network timeout error that is not important. A 404 is important but a network timeout I don't even need to really care about that. Different errors need to be handled differently. 

I believe this can be resolved using the `OnlineRepository` `FetchResponse` object. Have 2 failure states. You, the dev, decides what is important and what is not. A network timeout, you can say it's a failure but not a big deal (or even say to simply ignore the network response all together and have `OnlineRepository` just say that a fetch is done performing) or mark an error as a big deal and must be sticky.

**Comment by levibostian** [comment]
Another idea:

Dealing with the delivery of `DataState` to the UI. 

What if I did want to share the same UI for all of the different states of my data. I would ask `DataState`, "Give me the 1 important state of my data right now". If there is an error (sticky error or not sticky if I decide to go that route outlined in my comment above), then deliver the error as the 1 state the data is in. If it's empty and fetching, give me that state. etc. Give me 1 callback for the state so that I can determine the most important state my data is in. 

This idea can be summed up by saying, "If an error occurs, give me that error continuously until I mark it as resolved. Then, give me the other status of my data". After an error is marked as resolved, I could immediately deliver the state of the data (empty or not empty) to the user. 

Fallbacks to this idea is that if there is an error but fetching fresh data, I will want to show a UI along with the error saying I am fetching fresh data.

**Comment by levibostian** [comment]
After [this commit](https://github.com/levibostian/Teller-Android/commit/40fb0a17bf99354dfec3dc808179965d80bc1a8d), this issue has been somewhat resolved. The `error()` listener callback is called after `isEmpty()` for `OnlineDataState.deliver()`. This now allows you to use 1 view for all of your data states if you wish. However, the error being shown can be dismissed at anytime that the `OnlineDataState` is updated because the error is set to `null` after it's first delivered.

I still *may* implement some sort of "sticky" errors to make sure that errors do not go unnoticed. I also do like the idea of adding an "ignore" functionality to `FetchResult` in `OnlineRepository`. After these ideas are implemented, this issue may be resolved.

**Comment by levibostian** [comment]
From [this commit](https://github.com/levibostian/Teller-Android/commit/923be7cb7c142901280304c70953b7df5f1c35c2), I decided not to add "sticky" errors to the library. I do not believe it's needed as I will instead put the responsibility on the dev. Especially in this scenario: When either of the error callbacks get called in `OnlineDataStateListener` you decide to show an `AlertDialog`. If you call it multiple times, you will have multiple instances of `AlertDialog`s being shown in the UI instead of just once. I feel delivering the error to the dev once and then having them be in charge of it from there is the best way to handle this. It's simply (1) not needed for Teller to keep errors "sticky" as the dev can manage the error once it's delivered to the listener, (2) there is a lot more complexity added to the library that doesn't need to exist and (3) there are bad side effects from calling either of the error callbacks multiple times. 

Also, in the same commit I decided to no longer call `empty()` callback in `OnlineDataStateListener` if there was an error in the first fetch of data. I put a comment in the code to outline best practices instead. This helps to keep flexibility and remove the chance of any side effects such as this scenario: When the cache is empty, I want to show a UI that says, "You have no repos" for a GitHub repo listing app if there is no data in the cache. If there is an error during the first fetch of data in my GitHub app, I don't want to show that cache empty UI. I want to show a better message such as showing a "Try again" button that when pressed, I try fetching data again.
